dask_expr/_shuffle.py

```diff
@@ -618,7 +618,25 @@ def operation(cls, df, index, name: str, npartitions: int):
         return df.assign(**{name: index})
 
 
-class SetIndex(Expr):
+class BaseSetIndexSortValues(Expr):
+    def _divisions(self):
+        if self.user_divisions is not None:
+            return self.user_divisions
+        divisions, mins, maxes, presorted = _calculate_divisions(
```

Comment:
This function call can be expensive, and we are calling it in both `BaseSetIndexSortValues._divisions` and `SortValues._lower`. I know you are planning on doing some refactoring in a follow-up, but I would consider it a high priority to leverage some kind of caching here.

Reply:
**Another Note**: `_calculate_divisions` also requires the eager execution of a task graph, but doesn't leave us an opportunity to apply fusion first. We probably want to find a way to allow the user to configure the optimization options.

Reply:
The function is already cached. This shouldn’t be executed more than once 

yeah I agree this needs some refactoring. I want to avoid executing this in general it it’s not necessary when looking at the expression tree

Reply:
>_calculate_division

Ah okay, I didn't see the `lru_cache`. I look forward to the refactor, but agree we can do it in a follow-up as long as we are caching.

Reply:
Yeah caching on the class level didn't really work, that's why it's in a method